File: source/javascript/linked_list_reverse.js
function makeNode(val, next) { return { val, next }; }

function buildList(arr) {
    let head = null;
    for (let i = arr.length - 1; i >= 0; i--) head = makeNode(arr[i], head);
    return head;
}

function printList(head, label) {
    const parts = [];
    let curr = head;
    while (curr !== null) { parts.push(curr.val); curr = curr.next; }
    console.log(`${label}: ${parts.join(" ")}`);
}

function reverseList(head) {
    let prev = null, curr = head;
    while (curr !== null) {
        const next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}

let head = buildList([1, 2, 3, 4, 5]);
printList(head, "original");
const reversed = reverseList(head);
printList(reversed, "reversed");

File: source/python/linked_list_reverse.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next

def build_list(arr):
    head = None
    for x in reversed(arr):
        head = Node(x, head)
    return head

def print_list(head, label):
    parts = []
    curr = head
    while curr:
        parts.append(str(curr.val))
        curr = curr.next
    print(f"{label}: {' '.join(parts)}")

def reverse_list(head):
    prev = None
    curr = head
    while curr:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node
    return prev

head = build_list([1, 2, 3, 4, 5])
print_list(head, "original")
reversed_head = reverse_list(head)
print_list(reversed_head, "reversed")
